### PR TITLE
Add boolean/comparison operator support to ValueRanges

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95063
* #94985
* __->__ #94944
* #94939
* #94906

Pretty straightforward.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

File: torch/utils/_sympy/value_ranges.py [hunk 1]
```diff
@@ -175,6 +172,91 @@ def index_expr(self, index, dtype):
...
+        if a.lower == a.upper and b.lower == b.upper and a.lower == b.lower:
+            return ValueRanges.wrap(sympy.true)
+        elif a.lower > b.upper or b.lower > a.upper:  # ranges disjoint
+            return ValueRanges.wrap(sympy.false)
+        return ValueRanges(sympy.false, sympy.true)
+
+    @classmethod
+    def ne(cls, a, b):
+        a = ValueRanges.wrap(a)
+        b = ValueRanges.wrap(b)
```
Comment (lezcano) [inline torch/utils/_sympy/value_ranges.py:210]: nit. no need to wrap these.

File: torch/utils/_sympy/value_ranges.py [hunk 2]
```diff
@@ -175,6 +172,91 @@ def index_expr(self, index, dtype):
...
+
+    @classmethod
+    def ge(cls, a, b):
+        a = ValueRanges.wrap(a)
+        b = ValueRanges.wrap(b)
+        if a.lower >= b.upper:
+            return ValueRanges.wrap(sympy.true)
+        elif a.upper < b.lower:
+            return ValueRanges.wrap(sympy.false)
+        return ValueRanges(sympy.false, sympy.true)
```
Comment (lezcano) [inline torch/utils/_sympy/value_ranges.py:251]: You can implement these two in terms of not and lt / gt

**Review by lezcano — APPROVED** [review]
Very neat idea!

File: torch/utils/_sympy/value_ranges.py [hunk 3]
```diff
@@ -69,6 +69,11 @@ def __init__(self, lower, upper):
 
     def __contains__(self, x):
         x = simple_sympify(x)
+        if isinstance(x, SympyBoolean):
+            if x is sympy.true:
+                return self.upper == sympy.true
```
Comment (eellison) [inline torch/utils/_sympy/value_ranges.py:74]: `sympy.True in ValueRanges(-5, 5) -> False`. Could we change `if x is sympy.true` -> `if x is sympy.true and isinstance(self.lower, SympyBoolean) and isinstance(self.upper, SympyBoolean)`

File: torch/utils/_sympy/value_ranges.py [hunk 4]
```diff
@@ -175,6 +172,91 @@ def index_expr(self, index, dtype):
...
+        elif a.lower == a.upper and b.lower == b.upper:
+            return ValueRanges.wrap(sympy.And(a.lower, b.lower))
+        else:
+            return ValueRanges(sympy.false, sympy.true)
+
+    @staticmethod
+    def eq(a, b):
+        a = ValueRanges.wrap(a)
+        b = ValueRanges.wrap(b)
+        if a.lower == a.upper and b.lower == b.upper and a.lower == b.lower:
```
Comment (eellison) [inline torch/utils/_sympy/value_ranges.py:201]: `ValueRanges.unknown() == ValueRanges.unknown() => True`. All of these functions are not handling infs correctly.

**Review by eellison — COMMENTED** [review]
Generally looks good but isnt handling `ValueRanges.unknown()` correctly.
```
>>> sympy.oo == sympy.oo
True
```

We should fix that or we should check that the inputs are booleans before trying to do precise analysis.

**Comment by ezyang** [comment]
Err, I am confused. Unknown is -oo, oo; and oo != -oo. Am I missing something?

**Comment by ezyang** [comment]
Nyeh tests are failing though, I will look

**Review by eellison — COMMENTED** [review]
actually yea, you are right about the inf being okay. we still might need better sanity / reasoning on non bool inputs.

File: torch/utils/_sympy/value_ranges.py (hunk 4 above)
Comment (ezyang) [reply torch/utils/_sympy/value_ranges.py:208]: Just to reiterate, on the code on this PR:

```
>>> ValueRangeAnalysis.eq(ValueRanges.unknown(), ValueRanges.unknown())
ValueRanges(lower=False, upper=True)
```

which is correct.

**Comment by ezyang** [comment]
> we still might need better sanity / reasoning on non bool inputs.

Yeah, so the main footgun is you can't actually use unknown() for bool ranges. This seems fine, I'll beef up asserts for mixing bool/non-bool ranges.

**Pull request closed without merging**